src/services/accounts/AccountsAssetsService.ts

```diff
@@ -0,0 +1,141 @@
+import { ApiPromise } from '@polkadot/api';
+import { StorageKey } from '@polkadot/types';
+import { AssetBalance, AssetId, BlockHash } from '@polkadot/types/interfaces';
+import { AccountId } from '@polkadot/types/interfaces/runtime';
+import { AnyNumber } from '@polkadot/types/types';
+
+import {
+	IAccountAssetApproval,
+	IAccountAssetsBalances,
+	IAssetBalance,
+} from '../../types/responses';
+import { AbstractService } from '../AbstractService';
+
+export class AccountsAssetsService extends AbstractService {
+	/**
+	 * Fetch all the `AssetBalance`s along side their `AssetId`'s for a given array of queried AssetId`'s.
+	 * If none are queried the function will get all AssetId's associated with the
+	 * given `AccountId`, and send back all the `AssetsBalance`s.
+	 *
+	 * @param hash `BlockHash` to make call at
+	 * @param address `AccountId` associated with the balances
+	 * @param assets An array of `assetId`'s to be queried. If the length is zero
+	 * all assetId's associated to the account will be queried
+	 */
+	async fetchAssetBalances(
+		hash: BlockHash,
+		address: string,
+		assets: number[]
+	): Promise<IAccountAssetsBalances> {
+		const { api } = this;
+
+		const { number } = await api.rpc.chain.getHeader(hash);
+
+		let response;
+		if (assets.length === 0) {
+			/**
+			 * This will query all assets and return them in an array
+			 */
+			const keys = await api.query.assets.account.keysAt(hash);
```

Comment:
I don't fully know how this works but I am worried that this 1) be a huge query and 2) a lot to loop over. If we just want all the available assetIds why not just get the keys of [Asset](https://github.com/paritytech/substrate/blob/462653636821f542fd99beed25e817b7f14c38c9/frame/assets/src/lib.rs#L218-L222) or even [Metadata](https://github.com/paritytech/substrate/blob/462653636821f542fd99beed25e817b7f14c38c9/frame/assets/src/lib.rs#L252-L257)

Reply:
This is actually the same way that polkadot-js-ui queries all assets, thats where I got the idea from. I think it could definitely be costly. But I am not sure the above example with `Asset` and `Metadata` are correct. I might be wrong but that would only be for a single asset. 

Here is where and how its implemented in Polkadot-js apps. [useMapKeys](https://github.com/polkadot-js/apps/blob/master/packages/react-hooks/src/useMapKeys.ts#L16) is the hook that calls the `keysAt`, and then [useMapKeys is called here](https://github.com/polkadot-js/apps/blob/master/packages/page-assets/src/useAssetIds.ts#L19) inside of `useAssetIds`.

Reply:
I see what you are saying, your original logic mimics the app exactly afaict. I just don't understand why thats correct.

From my understanding we need to:
1) Get all `assetIds`.
2) Check every single `assetId` and see if the user has them.

Every `assetId` is stored in `Asset`. If you look at the `create` call you can [see the `assetId` is always inserted into `Assets`](https://github.com/paritytech/substrate/blob/462653636821f542fd99beed25e817b7f14c38c9/frame/assets/src/lib.rs#L381-L396).

Operating under the assumption that `keys()` will return all the combination of keys on a double map and not just the first key: in a single map I would expect that number of keys to be way less (#keys == #assets). On the other hand `Account` is a double map with 2 keys per entry; so the total of amount of keys could be much more and is always guaranteed to be a super set of the keys in `Account` (#keys == #assets * #accountIds)

Reply:
Yup you are 100% right. I changed the call to be `api.query.assets.asset.keysAt(hash)` which will only go through a Single Map StorageKey. Good catch and thanks for pointing that out.

Reply:
Resolved thanks!